### PR TITLE
conformance 1.7.20 4

### PR DESCRIPTION
**Pull request by abergs**
This is based on #456 and results in a 100% pass on the v1.7.20-4 tool.

image

- **FIDO Conformance Tools v1.7.15 fixes**
- **Json serialization fix**
- **Unit test fix**
- **tokenbindig, AppId, UVP**
- **unit test fix (tokenbinding dto parsing)**
- **fix azure pipeline**
- **Improve trustanchor test coverage**
- **TestPackedttestationAsyncFailTrustAnchorOnRootCertInTrustPath only works on Windows**
- **Do not make this private**
- **Keep Tokenbinding around**
- **Update AuthenticatorAssertionResponse.cs**
- **Added XML comments to requestTokenBinding**

### CODE REVIEW

File: Src/Fido2.Models/CredentialCreateOptions.cs
Comment (abergs) [inline Src/Fido2.Models/CredentialCreateOptions.cs:139]: P-8 Send a valid ServerAuthenticatorAttestationResponse with SELF "packed" attestation, for "ALG_SIGN_RSASSA_PKCSV15_SHA1_RAW" aka "RS1" algorithm, and check that server succeeds‣

File: Src/Fido2.Models/Objects/AuthenticationExtensionsClientInputs.cs
Comment (abergs) [inline Src/Fido2.Models/Objects/AuthenticationExtensionsClientInputs.cs:15]: JSON Serialization fails if this is object.

File: Src/Fido2.Models/Objects/AuthenticationExtensionsClientInputs.cs
Comment (abergs) [inline Src/Fido2.Models/Objects/AuthenticationExtensionsClientInputs.cs:23]: CleanShot 2024-07-16 at 10 24 02

Comment (abergs) [reply Src/Fido2.Models/Objects/AuthenticationExtensionsClientInputs.cs:23]: Conformance tests do not want AppId to be returned.

File: Src/Fido2/AuthenticatorAssertionResponse.cs
Comment (abergs) [inline Src/Fido2/AuthenticatorAssertionResponse.cs:68]: Conformance requires requestTokenBinding to still be around, should be removed once the tool is updated.

File: Src/Fido2/Extensions/CryptoUtils.cs
Comment (abergs) [inline Src/Fido2/Extensions/CryptoUtils.cs:64]: @aseigler Is it really ok to *not* equality compare the path?

Without this we get: 
CleanShot 2024-07-16 at 11 23 02

I could throw in a `|| conformanceTesting`-related check

File: Src/Fido2/TrustAnchor.cs
Comment (abergs) [inline Src/Fido2/TrustAnchor.cs:35]: @aseigler Would appreciate your eyes here too, but to me it just looks like we have the addition of above if-block that is more strict than without it.

File: Test/CryptoUtilsTests.cs
Comment (abergs) [inline Test/CryptoUtilsTests.cs:70]: These test changes are also something I wonder about @aseigler

**Comment by abergs** [comment]
@aseigler See mentions. If I don't get your approval I might move those changes to a secondary PR wand wait for your review, while still landing some of these fixes.

File: Src/Fido2.Models/CredentialCreateOptions.cs
Comment (aseigler) [reply Src/Fido2.Models/CredentialCreateOptions.cs:137]: Ideally, this should be made configurable instead of being hard coded, with defaults excluding RS1 (IMHO).

File: Src/Fido2/Extensions/CryptoUtils.cs
Comment (aseigler) [reply Src/Fido2/Extensions/CryptoUtils.cs:64]: The check to see if the subject and the issuer are the same is based on logic described in https://datatracker.ietf.org/doc/html/rfc5280#section-3.2.  I thought we had already covered this before, where the batch certificate referenced in the metadata is not a root...I'll have to review this more thoroughly.

File: Src/Fido2.Models/CredentialCreateOptions.cs
Comment (abergs) [reply Src/Fido2.Models/CredentialCreateOptions.cs:137]: Tracked in #533, resolving this conversation.

File: Src/Fido2/IFido2.cs
Comment (iamcarbon) [inline Src/Fido2/IFido2.cs:31]: Consider making a new object for the parameters so we can continue to add or remove parameters in the future without a binary breaking change.

```csharp
public class MakeNewCredentialRequest
{    
    public required AuthenticatorAttestationRawResponse AttestationResponse { get; init; }
    
    public required CredentialCreateOptions OriginalOptions { get; init; }
    
    // for Fido2 2024 conformance testing only
    [EditorBrowsable(EditorBrowsableState.Never)] 
    public byte[]? RequestTokenBindingId { get; init; }
}
```

File: Src/Fido2/TrustAnchor.cs
Comment (iamcarbon) [inline Src/Fido2/TrustAnchor.cs:11]: consider:

```
public static void Verify(MetadataBLOBPayloadEntry? metadataEntry, X509Certificate2[] trustPath, FidoValidationMode validationMode =  FidoValidationMode.Default)
```

**Review by iamcarbon — CHANGES_REQUESTED** [review]

File: Src/Fido2/Extensions/CryptoUtils.cs
Comment (iamcarbon) [reply Src/Fido2/Extensions/CryptoUtils.cs:64]: @abergs I saw that we removed the check to make sure the trustPath's root subject is equal to the issuer. 

What does the conformance test use for the Subject and Issuer (where they don't match). Is there a case, in production, where these values would also not match? Are they completely different values, or just encoded differently?

File: Src/Fido2/TrustAnchor.cs
Comment (iamcarbon) [reply Src/Fido2/TrustAnchor.cs:35]: @abergs Are we able to reference the rule in https://datatracker.ietf.org/doc/html/rfc5280 that this new condition enforces?

File: Src/Fido2/Fido2.cs
Comment (abergs) [reply Src/Fido2/Fido2.cs:75]: That's an interesting alternative to get it less visible!

File: Src/Fido2/Fido2.cs
Comment (abergs) [reply Src/Fido2/Fido2.cs:120]: Also a good idea!

File: Src/Fido2/Extensions/CryptoUtils.cs
Comment (abergs) [reply Src/Fido2/Extensions/CryptoUtils.cs:52]: Good idea!

File: Src/Fido2/Extensions/CryptoUtils.cs
Comment (abergs) [reply Src/Fido2/Extensions/CryptoUtils.cs:64]: @iamcarbon Not sure - I haven't looked into it.

File: Test/CryptoUtilsTests.cs
Comment (abergs) [reply Test/CryptoUtilsTests.cs:70]: just a reminder to give these changes a review @aseigler :)

File: Src/Fido2/Extensions/CryptoUtils.cs
Comment (abergs) [reply Src/Fido2/Extensions/CryptoUtils.cs]: As part of #554 I reverted this change until we get more data on this or @aseigler can review. I will open another PR to bring these changes back.

File: Test/CryptoUtilsTests.cs
Comment (abergs) [reply Test/CryptoUtilsTests.cs]: This was also reverted in #554

File: Src/Fido2/TrustAnchor.cs
Comment (googyi) [reply Src/Fido2/TrustAnchor.cs:35]: https://github.com/passwordless-lib/fido2-net-lib/pull/456

TrustAnchor.cs : 32
This exact testcase was fixed with this (conformance tests v1.7.15):

Server-ServerAuthenticatorAttestationResponse-Resp-5 Test server processing "packed" FULL attestation F-10 Send ServerAuthenticatorAttestationResponse with FULL "packed" attestation, with attStmt.x5c containing full chain, and check that server returns an error https://datatracker.ietf.org/doc/html/rfc5280#section-6.1

File: Src/Fido2/TrustAnchor.cs
Comment (abergs) [reply Src/Fido2/TrustAnchor.cs]: Resolved in: https://github.com/passwordless-lib/fido2-net-lib/pull/531/commits/dfa6f72b20d923fa917e48fbd7eeed24752d3921

File: Src/Fido2/Extensions/CryptoUtils.cs
Comment (abergs) [reply Src/Fido2/Extensions/CryptoUtils.cs]: Resolved in Resolved in: https://github.com/passwordless-lib/fido2-net-lib/pull/531/commits/dfa6f72b20d923fa917e48fbd7eeed24752d3921

File: Src/Fido2/Fido2.cs
Comment (abergs) [reply Src/Fido2/Fido2.cs:75]: @iamcarbon While drafting an implementation on your feedback  I decided against it since it would pollute the CredentialCreateOptions; they are expected to get serialised and sent to the client and should contain the "request token id" value.

File: Src/Fido2/Fido2.cs
Comment (abergs) [reply Src/Fido2/Fido2.cs:120]: Yes, let's do this in a separate PR.

File: Src/Fido2/TrustAnchor.cs
Comment (abergs) [reply Src/Fido2/TrustAnchor.cs:35]: @googyi Thanks for chiming in. What part of the [datatracker.ietf.org/doc/html/rfc5280#section-6.1](https://datatracker.ietf.org/doc/html/rfc5280#section-6.1) is deciding checking the last entry of the trustpath to root certs?

Comment (googyi) [reply Src/Fido2/TrustAnchor.cs:35]: I think this 6.1 section was referenced in that testcase, I don't know which part describes this exactly. 
This looks almost like that: 

"When the trust anchor is provided in the form of a self-signed
   certificate, this self-signed certificate is not included as part of
   the prospective certification path."

   But I am not sure that this is a self-signed cert.

Comment (abergs) [reply Src/Fido2/TrustAnchor.cs:35]: 6.1 also mentions " A certificate MUST NOT appear more than once in a prospective
   certification path." which may be fixed by this?

**Pull request merged**
